tests/jaxpr_effects_test.py

```diff
@@ -175,14 +205,153 @@ def f(x):
 
 class EffectfulJaxprLoweringTest(jtu.JaxTestCase):
 
-  def test_cannot_lower_jaxpr_with_effects_in_hop(self):
+  def test_should_pass_tokens_into_ordered_effect(self):
+
+    def _effect_lowering(ctx, *, effect):
+      self.assertListEqual(list(ctx.tokens_in.effects()), ['foo'])
+      ctx.set_tokens_out(ctx.tokens_in)
+      return []
+    mlir.register_lowering(effect_p, _effect_lowering)
+
+    @jax.jit
+    def f(x):
+      effect_p.bind(effect='foo')
+      return x + 1.
+    f.lower(2.)
+
+  def test_should_not_pass_tokens_into_unordered_effect(self):
+
+    def effect_lowering(ctx, *, effect):
+      self.assertEmpty(ctx.tokens_in)
+      return []
+    mlir.register_lowering(effect_p, effect_lowering)
+
+    @jax.jit
+    def f(x):
+      effect_p.bind(effect='bar')
+      return x + 1.
+    f.lower(2.)
+
+  def test_lowering_that_doesnt_set_tokens_should_cause_error(self):
+
+    def bad_effect_lowering(ctx, *, effect):
+      # Doesn't call `ctx.set_tokens_out`!
+      return []
+    mlir.register_lowering(effect_p, bad_effect_lowering)
+
+    @jax.jit
+    def f(x):
+      effect_p.bind(effect='foo')
+      return x + 1.
+    with self.assertRaisesRegex(ValueError, 'Lowering rule for `effect` needs to '
+        'set `tokens_out`'):
+      f.lower(2.)
+
+  def test_lowering_that_sets_wrong_tokens_should_cause_error(self):
+
+    def bad_effect_lowering(ctx, *, effect):
+      ctx.set_tokens_out(mlir.TokenSet(bar=ctx.tokens_in.get('foo')))
+      return []
+    mlir.register_lowering(effect_p, bad_effect_lowering)
+
+    @jax.jit
+    def f(x):
+      effect_p.bind(effect='foo')
+      return x + 1.
+    with self.assertRaisesRegex(ValueError, 'Lowering rule for `effect` returns '
+        'incorrect set of output token.'):
+      f.lower(2.)
+
+  def test_lowering_ordered_effect_should_create_tokens(self):
```

Comment:
Another option for testing the MHLO that is generated would be to use a FileCheck test. I am not saying you must do this but it may be preferable and perhaps more readable.

See
https://llvm.org/docs/CommandGuide/FileCheck.html
and some of the existing tests for other properties:
https://github.com/google/jax/tree/main/tests/filecheck

(These tests aren't routinely tested in our OSS CI, though. We do test them internally in CI, mostly because they aren't things you can easily hook up to pytest but one can easily hook up to Bazel.)

The advantage of doing this is that you don't have to roll your own pattern matching code. You may or may not prefer that, depending on the pattern being matched.

Reply:
Thanks for the pointer, FileCheck seems super useful. I think the current tests are simple enough (patterns are not very advanced) but I'll work on some more advanced tests using filecheck as well. These I'll submit internally since it's easier to test that way as you mentioned.